src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetPowerShellUsageCollector.cs

```diff
@@ -0,0 +1,421 @@
+                    TelemetryActivity.EmitTelemetryEvent(_vsSolutionData.ToTelemetryEvent());
+                    IncrementUpdateVSInstanceData();
+                }
+
+                ClearSolutionData();
+                _vsSolutionData._solutionLoaded = true;
+            }
+        }
+
+        private void NuGetPowerShellUsage_SolutionCloseHandler()
+        {
+            lock (_lock)
+            {
+                // Emit solution telemetry
+                TelemetryActivity.EmitTelemetryEvent(_vsSolutionData.ToTelemetryEvent());
+                IncrementUpdateVSInstanceData();
+                ClearSolutionData();
+            }
+        }
+
+        private void NuGetPowerShellUsage_VSInstanseCloseHandler()
+        {
+            lock (_lock)
+            {
+                if (_powerShellLoadEvent != null)
+                {
+                    TelemetryActivity.EmitTelemetryEvent(_powerShellLoadEvent);
+                    _powerShellLoadEvent = null;
+                }
+
+                // Edge case: PMC used without solution load
+                if (!_vsSolutionData._solutionLoaded && _vsSolutionData._nuGetPMCExecuteCommandCount > 0)
+                {
+                    // PMC used before any solution is loaded, let's emit what we have for nugetvsinstanceclose event aggregation before loading a solution.
+                    TelemetryActivity.EmitTelemetryEvent(_vsSolutionData.ToTelemetryEvent());
+                    IncrementUpdateVSInstanceData();
+                }
+
+                // Emit VS Instance telemetry
+                TelemetryActivity.EmitTelemetryEvent(_vsInstanceData.ToTelemetryEvent());
+            }
+        }
+
+        //If open new solution then need to clear previous solution events. But powershell remain loaded in memory.
+        private void ClearSolutionData()
+        {
+            bool pmcPowershellLoad = _vsSolutionData._loadedFromPMC;
+            bool pmuiPowershellLoad = _vsSolutionData._loadedFromPMUI;
+
+            _vsSolutionData = new NuGetPowershellVSSolutionCloseEvent();
+
+            _vsSolutionData._loadedFromPMC = pmcPowershellLoad;
+            _vsSolutionData._loadedFromPMUI = pmuiPowershellLoad;
+        }
+
+        private void IncrementUpdateVSInstanceData()
+        {
+            _vsInstanceData._nugetPMCExecuteCommandCount += _vsSolutionData._nuGetPMCExecuteCommandCount;
+            _vsInstanceData._nugetPMCWindowLoadCount += _vsSolutionData._nuGetPMCWindowLoadCount;
+            _vsInstanceData._nugetPMUIExecuteCommandCount += _vsSolutionData._nuGetPMUIExecuteCommandCount;
+
+            if (_vsSolutionData._loadedFromPMC)
+            {
+                _vsInstanceData._pmcLoadedSolutionCount++;
+            }
+
+            if (_vsSolutionData._loadedFromPMUI)
+            {
+                _vsInstanceData._pmuiLoadedSolutionCount++;
```

Comment:
This method (`IncrementUpdateVSInstanceData`) is being called on both solution open and solution close. Therefore, this "solution count" will be wrong (hopefully always exactly double the real value).